File: app/src/main/java/com/example/mercury/nearbysample/MessageAdapter.java
package com.example.mercury.nearbysample;

import android.content.res.Resources;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MessageAdapter extends RecyclerView.Adapter<MessageAdapter.MessageViewHolder> {

    private static final String TAG = MessageAdapter.class.getSimpleName();
    private ArrayList<MessageModel> messageList;
    private MainActivityCallback activityCallback;

    MessageAdapter(ArrayList<MessageModel> messages, MainActivityCallback activityCallback) {
        messageList = messages;
        this.activityCallback = activityCallback;
    }

    @Override
    public MessageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.message_item, parent, false);
        return new MessageViewHolder(v);
    }


    @Override
    public void onBindViewHolder(MessageViewHolder holder, int position) {
        //TODO: color message based on sender
        MessageModel message = messageList.get(position);
        holder.messageWho.setText(message.getWho());
        holder.messageWho.setText(String.format("%s : ", message.getWho()));
        if (message.getWho().equals(activityCallback.getUsername())) {
            holder.messageWho.setTextColor(activityCallback.getResources().getColor(R.color.colorAccent));
            holder.messageWho.setBackgroundColor(activityCallback.getResources().getColor(R.color.selfMessageBackground));
        }
        holder.messageText.setText(message.getText());
    }

    @Override
    public int getItemCount() {
        return messageList.size();
    }

    public void setMessageList(ArrayList<MessageModel> newMessages) {
        messageList = newMessages;
        notifyDataSetChanged();
        Log.d(TAG, "setMessageList: updating adapter's knowledge of message list. messageList size: " + String.valueOf(messageList.size()));
    }

    class MessageViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.message_who)
        TextView messageWho;

        @BindView(R.id.message_text)
        TextView messageText;

        LinearLayout messageContainer;

        MessageViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            messageContainer = (LinearLayout) itemView;
        }

    }

    interface MainActivityCallback{
        String getUsername();
        Resources getResources();
    }
}

File: app/src/main/java/com/example/mercury/nearbysample/MessageAdapter.kt
package com.example.mercury.nearbysample

import android.content.res.Resources
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView

import java.util.ArrayList

import butterknife.BindView
import butterknife.ButterKnife
import com.example.mercury.nearbysample.R.id.message_text
import com.example.mercury.nearbysample.R.id.message_who
import com.example.mercury.nearbysample.R.layout.message_item
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.message_item.*
import kotlinx.android.synthetic.main.message_item.message_text



open class MessageAdapter constructor(private var messageList: ArrayList<MessageModel>,
                                      private val activityCallback: MainActivityCallback,
                                      override val containerView: View?)
    : RecyclerView.Adapter<MessageAdapter.MessageViewHolder>(), LayoutContainer {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MessageViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.message_item, parent, false)
        return MessageViewHolder(v)
    }


    override fun onBindViewHolder(holder: MessageViewHolder, position: Int) {
        //TODO: color message based on sender
        val message = messageList[position]
        if (message.who == activityCallback.username) {
            message_text.background = activityCallback.resources.getDrawable(R.drawable.self_message_background)
            holder.messageContainer.layoutDirection = View.LAYOUT_DIRECTION_RTL
            //these calls are necessary on older OS for some reason
            message_who.text = null
        } else {
            message_who.text = String.format("%s : ", message.who)
            //these calls are necessary on older OS for some reason
            message_who.setTextColor(activityCallback.resources.getColor(R.color.colorPrimary))
            message_text.background = null
            holder.messageContainer.layoutDirection = View.LAYOUT_DIRECTION_LTR
        }
        message_text.text = message.text
    }

    override fun getItemCount(): Int {
        return messageList.size
    }

    fun setMessageList(newMessages: ArrayList<MessageModel>) {
        messageList = newMessages
        notifyDataSetChanged()
        Log.d(TAG, "setMessageList: updating adapter's knowledge of message list. messageList size: " + messageList!!.size.toString())
    }

    inner class MessageViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var messageContainer: LinearLayout

        init {
            ButterKnife.bind(this, itemView)
            messageContainer = itemView as LinearLayout
        }

    }

    interface MainActivityCallback {
        val username: String
        val resources: Resources
    }

    companion object {

        private val TAG = MessageAdapter::class.java.simpleName
    }
}
